HPCC-11764 Dafilesrv authentication and encryption on transport

RussWhitehead opened the pull request:
Current dafilesrv uses unencrypted sockets which is not secure. This update
adds a new section to the environment.conf file to optionally specify that
dafilesrv and its clients should use SSL, and points to the cert file, key file
and provides an encrypted passphrase. If specified, dafilesrv and its clients
will enable secure sockets (openSSL) and communicate in a secure manner.

Signed-off-by: William Whitehead william.whitehead@lexisnexis.com

RussWhitehead committed:
HPCC-11764 Dafilesrv authentication and encryption on transport

Current dafilesrv uses unencrypted sockets which is not secure. This update
adds a new section to the environment.conf file to optionally specify that
dafilesrv and its clients should use SSL, and points to the cert and key
files.
If specified, dafilesrv and its clients
will enable secure sockets (openSSL) and communicate in a secure manner.

Signed-off-by: William Whitehead <william.whitehead@lexisnexis.com>

ghalliday committed:
HPCC-12663 Reduce the default page size, default packed on in thor

Also optimize finding a free page by using __builtin_ffs

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
HPCC-12665 Use a list to maintain a list of heaplets that may have space

* Add a lockless singly linked list of heaplets that may potentially
  have any spare space.  This avoids walking all the heaplets when checking
  if any existing blocks have space, and when checking for heaplets to
  return to the global memory pool.

* Convert the list of heaplets for a heap from singly to doubly linked to
  optimize removal of empty heaplets.

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

wangkx committed:
HPCC-11714 Log soap request type/user/ip in a line

The existing ESP code logs soap request type/user/ip
in two lines. This fix merges the two log lines into
one line.

The existing code does log http request type/user/ip
in a line.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

mckellyln committed:
HPCC-12939 Add include file to make sync_file_region call more efficient

Signed-off-by: Mark Kelly <mark.kelly@lexisnexis.com>

jakesmith committed:
HPCC-9395 Expand hashjoin to hashdistributes+local join.

Also adjust resourcing of DISTRIBUTE, allowing for greater than 2.
At the moment, with the other hard limits, the practical limit is
increased to 4.

Signed-off-by: Jake Smith <jake.smith@lexisnexis.com>

ghalliday committed:
HPCC-11559 Don't implicitly create ordered appends from disk by default

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
HPCC-12665 Update comments and other minor changes following review

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

afishbeck committed:
HPCC-11482 Automatically add manifests named same as included ECL files

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

Commit:
HPCC-12228 DOCS:Add ver_ info to SOAPCALL

Signed-off-by: Jim DeFabia <jamesdefabia@lexisnexis.com>

jakesmith committed:
HPCC-12915 Reset send queue, on lookup/smart restart (in loop)

The send block queue was not being reset when lookup/smart join
was reran (inside a loop). As a consequence a terminator
block (NULL) may have still be left on the queue and caused the
row processor to terminate early and cause deadlock on the
producer side.

Signed-off-by: Jake Smith <jake.smith@lexisnexis.com>

Commit:
HPCC-12948 Add DojoD32DChart to copy only list

Fixes HPCC-12948

Signed-off-by: Gordon Smith <gordon.smith@lexisnexis.com>

Commit:
HPCC-12809 Landing Zone Spray Group shows too many clusters

Fixes HPCC-12809

Signed-off-by: Gordon Smith <gordon.smith@lexisnexis.com>

richardkchapman committed:
Merge pull request #6786 from ghalliday/issue12665

HPCC-12665 Use a list to maintain a list of heaplets that may have space

Reviewed-By: Mark Kelly <mark.kelly@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #6913 from GordonSmith/HPCC-12948

HPCC-12948 Add DojoD32DChart to copy only list

Reviewed-By: Xiaoming Wang <xiaoming.wang@lexisnexis.com>

richardkchapman committed:
Merge pull request #6910 from JamesDeFabia/12228Ver

HPCC-12228 DOCS:Add ver_ info to SOAPCALL

Reviewed-By: Kevin Wang <kevin.wang@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

Commit:
HPCC-12951 Refresh WU Lists on Tab Selection

Fixes HPCC-12951

Signed-off-by: Gordon Smith <gordon.smith@lexisnexis.com>

richardkchapman committed:
Merge pull request #6908 from ghalliday/issue11559

HPCC-11559 Don't implicitly create ordered appends from disk by default

Reviewed-By: Jake Smith <jake.smith@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #6707 from ghalliday/issue12663

HPCC-12663 Reduce the default page size, default packed on in thor

Reviewed-By: Mark Kelly <mark.kelly@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #6906 from jakesmith/hpcc-9395

HPCC-9395 Expand hashjoin to hashdistributes+local join.

Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #6918 from jakesmith/hpcc-12915

HPCC-12915 Reset send queue, on lookup/smart restart (in loop)

Reviewed-By: Gavin Halliday <gavin.halliday@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge branch 'candidate-5.0.6' into candidate-5.2.0

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #6904 from mckellyln/hpcc12939

HPCC-12939 Added include file for sync_file_region call to be more efficient

Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

jakesmith commented:
@RussWhitehead - can you look at my comments from 24th Nov.

AttilaVamos committed:
HPCC-12863 Option to pass options from xml to a query

Add code to automatically check and process xml file to generate
stored variables.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

mckellyln commented on system/jlib/jsocket.cpp:
```diff
@@ -6015,3 +6015,96 @@ ISocketConnectWait *nonBlockingConnect(SocketEndpoint &ep,unsigned connecttimeou
+#endif
+                readySocks.append(socks.item(idx));
+                if (readySocks.length() == res)
+                    break;
```
Probably never happen, but maybe put a max trip count on this loop with idx <= socks.length() ?

RussWhitehead replied:
agreed, but i think should be    idx < socks.length()    ?

mckellyln commented on system/jlib/jutil.cpp:
```diff
@@ -2417,6 +2419,62 @@ IPropertyTree *getHPCCEnvironment(const char *configFileName)
+                Owned<IProperties> conf = createProperties(configFileSpec.str(), true);
+                useSSL = conf->getPropBool("dfsUseSSL", false);
+                certificate.set(conf->queryProp("dfsSSLCertFile"));
+                privateKey.set(conf->queryProp("dfsSSLPrivateKeyFile"));
```
Should we always set certificate and privateKey if dfsUseSSL is false ?

RussWhitehead replied:
I agree

mckellyln commented on common/remote/sockfile.cpp:
```diff
@@ -4365,12 +4431,67 @@ class CRemoteFileServer : public CInterface, implements IRemoteFileServer, imple
+                {
+                    UnsignedArray waitingSocks;
+                    //SSL Enabled. Listen for non SSL connection on DAFILESRV_PORT and reject them
+                    int numReady = wait_read_multiple(readSocks, 1000*60*1, waitingSocks);
```
I think using select() for 2 fds is fine.

RussWhitehead replied:
Leaving the code as is (using select() for 2 fds) unless someone protests

richardkchapman committed:
Merge pull request #6919 from AttilaVamos/HPCC-12863-improvement

HPCC-12863 Option to pass options from xml to a query

Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

Commit:
HPCC-12940 DOCS:Update VM doc for a VMPlayer image

FIX HPCC-12940 DOCS:Update VM doc for a VMPlayer image
Update Documenation to reflect that there will be multiple
VM Images available, one for VirtualBox and one for VMPlayer

Signed-off-by: G Panagiotatos <g.pan3.14@gmail.com>

richardkchapman committed:
Merge pull request #6920 from g-pan/H12940-VMErr

HPCC-12940 DOCS:Update VM doc for a VMPlayer image

Reviewed-By:Richard Chapman <rchapman@hpccsystems.com>

mckellyln commented on common/remote/sockfile.cpp:
```diff
@@ -4365,12 +4431,67 @@ class CRemoteFileServer : public CInterface, implements IRemoteFileServer, imple
+            //Create unsecure socket to reject non-ssl client requests
+            StringBuffer ip;
+            IpAddress ipAddr;
+            acceptsock->getPeerAddress(ipAddr);
```
Are we connected to anything at this point ?
I get "getpeername failed 107" error msg

richardkchapman committed:
Merge pull request #6916 from afishbeck/jsonBigNumberAsString

HPCC-12785 Output JSON BigIntegers as strings

Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

mckellyln commented:
Note the passphrase must be removed from the key or else the program will fail with a bad password read error.  Is an uncrypted key file ok ?

richardkchapman committed:
Merge pull request #6911 from afishbeck/autoManifest4

HPCC-11482 Automatically add manifests named same as included ECL files

Reviewed-By: Gordon Smith <gordon.smith@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #6887 from wangkx/h11714

HPCC-11714 Log soap request type/user/ip in a line

Reviewed-By: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #6917 from GordonSmith/HPCC-12951

HPCC-12951 Refresh WU Lists on Tab Selection

Reviewed-By: Miguel Vazquez <miguel.vazquez@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #6915 from GordonSmith/HPCC-12809

HPCC-12809 Landing Zone Spray Group shows too many clusters

Reviewed-By: Miguel Vazquez <miguel.vazquez@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

RussWhitehead committed:
Signed-off-by: William Whitehead <william.whitehead@lexisnexis.com>

RussWhitehead commented:
Closing, and replacing with https://github.com/hpcc-systems/HPCC-Platform/pull/6921
